### PR TITLE
add pipeline trigger syntax and expression

### PR DESCRIPTION
**Pull request by pipipipipi43**
#### What type of this PR

Add one of the following kinds:
/kind feature

#### What this PR does / why we need it:

Added pipeline trigger syntax and expression

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://terminus-org.app.terminus.io/erda/dop/projects/387/issues/all?id=208384&issueFilter__urlQuery=eyJzdGF0ZUJlbG9uZ3MiOlsiT1BFTiIsIldPUktJTkciLCJXT05URklYIiwiUkVPUEVOIl0sImFzc2lnbmVlSURzIjpbIjEwMDA4MDUiXX0%3D&issueTable__urlQuery=eyJwYWdlTm8iOjF9&issueViewGroup__urlQuery=eyJ2YWx1ZSI6ImthbmJhbiIsImNoaWxkcmVuVmFsdWUiOnsia2FuYmFuIjoiZGVhZGxpbmUifX0%3D&iterationID=506&type=TASK)
- [test-link](https://erda.dev.terminus.io/terminus/dop/projects/1/apps/3/pipeline?nodeId=MS8zL3RyZWUvZGV2ZWxvcC9waXBlbGluZS55bWw%3D&pipelineID=34481)

### CODE REVIEW

File: apistructs/pipeline_yml.go [hunk 1]
```diff
@@ -204,3 +204,8 @@ type PipelineYmlParseGraphResponse struct {
 	Header
 	Data *PipelineYml `json:"data"`
 }
+
+type PipelineTrigger struct {
+	EventName string            `yaml:"eventName,omitempty" json:"eventName,omitempty"`
+	Filter    map[string]string `yaml:"filter,omitempty" json:"filter,omitempty"`
```
Comment (Effet) [inline apistructs/pipeline_yml.go:210]: I think `Event` is better than`EventName`.

Comment (Effet) [reply apistructs/pipeline_yml.go:210]: And why Filter is a map, can you describe an example?

**Review by Effet — DISMISSED** [review]

File: pkg/parser/pipelineyml/samples/trigger.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+version: "1.1"
```
Comment (sfwn) [inline pkg/parser/pipelineyml/samples/trigger.yml:1]: adjust to latest yaml grammar.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: apistructs/pipeline_yml.go [hunk 3]
```diff
@@ -59,9 +59,9 @@ type PipelineYml struct {
 	// 2) 当 needUpgrade 为 false 时：
 	//    1) 用户传入的为 YAML(apistructs.PipelineYml) 时，ymlContent 返回 YAML(spec.PipelineYml)
 	//    2) 用户传入的为 YAML(spec.PipelineYml) 时，返回优化后的 YAML(spec.PipelineYml)
-	YmlContent string         `json:"ymlContent,omitempty"`
-	On         *TriggerConfig `json:"on,omitempty"`
-
+	YmlContent string             `json:"ymlContent,omitempty"`
+	On         *TriggerConfig     `json:"on,omitempty"`
+	Trigger    []*PipelineTrigger `json:"triggers,omitempty"`
```
Comment (kakj-go) [inline apistructs/pipeline_yml.go:64]: Trigger => Triggers

File: modules/pipeline/precheck/precheck.go [hunk 4]
```diff
@@ -74,6 +74,13 @@ func PreCheck(ctx context.Context, pipelineYmlByte []byte, itemForCheck precheck
 		return true, showMessage
 	}
 
+	triggerLables := make(map[string]map[string]string)
+	if y.Spec().Trigger != nil {
+		for _, val := range y.Spec().Trigger {
```
Comment (kakj-go) [inline modules/pipeline/precheck/precheck.go:79]: same like
```golang
if event.On != "" && len(event.Filter) > 0 {
    triggerLabels[event.On] = event.Filter
}
```

File: modules/pipeline/precheck/precheck.go [hunk 5]
```diff
@@ -74,6 +74,13 @@ func PreCheck(ctx context.Context, pipelineYmlByte []byte, itemForCheck precheck
...
+			triggerLables[val.On] = val.Filter
+		}
+	}
+	fmt.Println("init preCheck triggerLables: ", triggerLables)
```
Comment (kakj-go) [inline modules/pipeline/precheck/precheck.go:83]: remove

File: pkg/parser/pipelineyml/pipelineyml.go [hunk 6]
```diff
@@ -48,6 +48,9 @@ type PipelineYml struct {
 	needUpgrade        bool   // 是否需要升级
 	upgradedYmlContent []byte // 升级后的 yml content
 
+	// triggerLabel
```
Comment (sfwn) [inline pkg/parser/pipelineyml/pipelineyml.go:51]: triggerLabels xxx

File: pkg/parser/pipelineyml/samples/trigger.yml [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+triggers:
+  - "on": issue
```
Comment (sfwn) [inline pkg/parser/pipelineyml/samples/trigger.yml:3]: ugly quote around `on`.

File: pkg/parser/pipelineyml/visitor_labels.go [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+func NewLabelsVisitor(data []byte, labels map[string]string) *LabelsVisitor {
+	v := LabelsVisitor{}
+	v.data = data
+	v.labels = labels
+	return &v
+}
+
+// yaml 全局文本替换
```
Comment (sfwn) [inline pkg/parser/pipelineyml/visitor_labels.go:37]: english comment

**Review by sfwn — CHANGES_REQUESTED** [review]

File: pkg/parser/pipelineyml/pipelineyml.go (hunk 6 above)
Comment (pipipipipi43) [reply pkg/parser/pipelineyml/pipelineyml.go]: done

File: pkg/parser/pipelineyml/visitor_labels.go (hunk 8 above)
Comment (pipipipipi43) [reply pkg/parser/pipelineyml/visitor_labels.go]: done

File: pkg/parser/pipelineyml/samples/trigger.yml (hunk 7 above)
Comment (pipipipipi43) [reply pkg/parser/pipelineyml/samples/trigger.yml]: todo commit at  pipeline.Spec

**Review by sfwn — APPROVED** [review]

**Pull request merged**